### PR TITLE
Fix int overflow error when spilling large page

### PR DESCRIPTION
**Pull request by rschlussel**
Page serialization requires page size to fit in an integer, so larger
pages could hit an int overflow error.

Test plan - Not sure how to write a test because if I create a page that big I run out of java heapspace.  I plan to verify with a production query that encountered this issue to check if it fixes it, but need to wait for other testing to finish on that cluster first.

```
== RELEASE NOTES ==

General Changes
* Fix an issue where queries could encounter an int overflow error during spilling

```

### CODE REVIEW

**Review by sachdevs — COMMENTED** [review]
Nice one, I had recently started seeing this in my tests as well.

Was this a join spilling query or aggregation? Are we concerned about memory usage in `page.getRegion()` during splitPage()? I wonder if we are temporarily doubling memory usage by doing the page split.

File: presto-main/src/main/java/com/facebook/presto/spiller/FileSingleStreamSpiller.java [hunk 1]
```diff
@@ -144,11 +144,15 @@ private void writePages(Iterator<Page> pageIterator)
             while (pageIterator.hasNext()) {
                 Page page = pageIterator.next();
                 spilledPagesInMemorySize += page.getSizeInBytes();
-                SerializedPage serializedPage = serde.serialize(page);
-                long pageSize = serializedPage.getSizeInBytes();
-                localSpillContext.updateBytes(pageSize);
-                spillerStats.addToTotalSpilledBytes(pageSize);
-                writeSerializedPage(output, serializedPage);
+                // page serialization requires  page.getSizeInBytes() + Integer.BYTES to fit in an integer
+                splitPage(page, Integer.MAX_VALUE - Integer.BYTES).stream()
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/spiller/FileSingleStreamSpiller.java:148]: I think we can split with much moderate size? -- e.g. something like 8M?

**Review by wenleix — APPROVED** [review]
LGTM. I guess I also need to "backport" it to `TempStorageSingleStreamSpiller` 😂

**Comment by rschlussel** [comment]
> Was this a join spilling query or aggregation? 
Query had multiple joins and aggregations, so I'm not sure which was the culprit.

> Are we concerned about memory usage in `page.getRegion()` during splitPage()? I wonder if we are temporarily doubling memory usage by doing the page split.

That's a great question.  I think you're right.  I'm also not sure that memory is accounted for anywhere. I wonder if it can cause GC issues for the cluster (not just for spill, but anytime we use splitPage()).

**Review by sachdevs — APPROVED** [review]
Yeah, in my experience testing on T10s, this kinda stuff can cause JVM OOMs since in that one momentary instance the memory doubles and it's not yielding for memory anywhere. I'm good for merging this for now and then deal with the problem when it comes up though. I believe the solution will just be to try to make smaller pages when spilling!

File: presto-main/src/main/java/com/facebook/presto/spiller/FileSingleStreamSpiller.java [hunk 2]
```diff
@@ -144,11 +145,15 @@ private void writePages(Iterator<Page> pageIterator)
             while (pageIterator.hasNext()) {
                 Page page = pageIterator.next();
                 spilledPagesInMemorySize += page.getSizeInBytes();
-                SerializedPage serializedPage = serde.serialize(page);
-                long pageSize = serializedPage.getSizeInBytes();
-                localSpillContext.updateBytes(pageSize);
-                spillerStats.addToTotalSpilledBytes(pageSize);
-                writeSerializedPage(output, serializedPage);
+                // page serialization requires  page.getSizeInBytes() + Integer.BYTES to fit in an integer
+                splitPage(page, DEFAULT_MAX_PAGE_SIZE_IN_BYTES).stream()
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/spiller/FileSingleStreamSpiller.java:149]: We should probably add a similar fix to the storage based spiller

CC: @wenleix

**Review by arhimondr — APPROVED** [review]
Interesting. How are we ending up with 2GB+ pages?

**Comment by rschlussel** [comment]
> Interesting. How are we ending up with 2GB+ pages?

I'm not sure.   Follow up task is to look into that and prevent whatever is causing it from generating such large pages.

**Pull request merged**